staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -674,7 +675,221 @@ type crdConversionRESTOptionsGetter struct {
 	if err == nil {
-		ret.StorageConfig.Codec = versioning.NewCodec(ret.StorageConfig.Codec, ret.StorageConfig.Codec, t.converter, &unstructuredCreator{}, discovery.NewUnstructuredObjectTyper(), &unstructuredDefaulter{delegate: Scheme}, t.encoderVersion, t.decoderVersion)
+		d := validatingDecoder{delegate: ret.StorageConfig.Codec, validator: unstructuredSchemaValidator{
+			// drop invalid fields while decoding old CRs (before we had any ObjectMeta validation)
+			dropInvalidMetadata: true,
+		}}
+		c := validatingConverter{delegate: t.converter, validator: unstructuredSchemaValidator{}}
+		ret.StorageConfig.Codec = versioning.NewCodec(ret.StorageConfig.Codec, d, c, &unstructuredCreator{}, discovery.NewUnstructuredObjectTyper(), &unstructuredDefaulter{delegate: Scheme}, t.encoderVersion, t.decoderVersion)
 	}
 	return ret, err
 }
+
+// validatingDecoder calls the delegate decoder, and then applies the Unstructured validator.
+type validatingDecoder struct {
+	delegate  runtime.Decoder
+	validator unstructuredSchemaValidator
+}
+
+var _ runtime.Decoder = validatingDecoder{}
+
+func (d validatingDecoder) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+	obj, gvk, err := d.delegate.Decode(data, defaults, into)
+	if err != nil {
+		return nil, gvk, err
+	}
+	if u, ok := obj.(*unstructured.Unstructured); ok {
+		if err := d.validator.apply(u); err != nil {
+			return nil, gvk, err
+		}
+	}
+
+	return obj, gvk, nil
+}
+
+// validatingConverter calls the delegate converter and applies the Unstructured validator.
+type validatingConverter struct {
+	delegate  runtime.ObjectConvertor
+	validator unstructuredSchemaValidator
+}
+
+var _ runtime.ObjectConvertor = validatingConverter{}
+
+func (v validatingConverter) Convert(in, out, context interface{}) error {
+	if err := v.delegate.Convert(in, out, context); err != nil {
+		return err
+	}
+
+	if u, ok := out.(*unstructured.Unstructured); ok {
+		if err := v.validator.apply(u); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v validatingConverter) ConvertToVersion(in runtime.Object, gv runtime.GroupVersioner) (runtime.Object, error) {
+	out, err := v.delegate.ConvertToVersion(in, gv)
+	if err != nil {
+		return nil, err
+	}
+
+	if u, ok := out.(*unstructured.Unstructured); ok {
+		if err := v.validator.apply(u); err != nil {
+			return nil, err
+		}
+	}
+
+	return out, nil
+}
+
+func (v validatingConverter) ConvertFieldLabel(version, kind, label, value string) (string, string, error) {
+	return v.ConvertFieldLabel(version, kind, label, value)
+}
+
+// unstructuredSchemaValidator does the validation for Unstructured that json.Unmarshal
+// does for native types. This includes:
+// - validating and pruning ObjectMeta (here with optional error instead of pruning)
+// - TODO: application of an OpenAPI validator (against the whole object or a top-level field of it).
```

Comment:
Why would my decoder perform the validation?  That seems odd.  Pruning and defaulting make sense to me, but validating during decoding seems very likely to fail.

Reply:
validating in the sense of schema validating. The same is done by `json.Unmarshal`.

Reply:
I think just the naming is tripping me up. `coercingDecoder` makes it clearer that schema coercion is the purpose. The option to drop data that can't be coerced (which we'd do when reading from etcd) or to fail on data that can't be coerced (which we'd do when accepting API requests) makes sense.

Reply:
> I think just the naming is tripping me up. coercingDecoder makes it clearer that schema coercion is the purpose.

Yeah, that would do for me.  I just don't want to see this as a spot to inject real validation checks.  Validating on reads (which I think this does) is dangerous.